package/gluon-mesh-vpn-core/files/lib/netifd/proto/gluon_wireguard.sh

```diff
@@ -0,0 +1,86 @@
+#!/bin/sh
+
+PROTO_DEBUG=1
+
+. /lib/functions.sh
+. ../netifd-proto.sh
+init_proto "$@"
+
+WG=/usr/bin/wg
+
+proto_gluon_wireguard_init_config() {
+	#no_device=1
+	#available=1
+	proto_config_add_int index
+	proto_config_add_int mtu
+}
+
+interface_linklocal_from_wg_public_key() {
+	# We generate a predictable v6 address
+	local macaddr="$(printf "%s" "$1"|md5sum|sed 's/^\(..\)\(..\)\(..\)\(..\)\(..\).*$/02:\1:\2:\3:\4:\5/')"
```

Comment:
The `md5sum` here feels extremely weird to me. Could the first bytes of the public key be used directly, without hashing, after decoding the base64? Are they distributed uniformly?

Reply:
This is also used on the server side to generate a predictable v6 address so I would refrain from changing this.

Reply:
https://github.com/freifunkMUC/wgkex/blob/6cd3db50d2c133eea818a3631108cf676b121f17/wgkex/worker/netlink.py#L39

For example.

Reply:
Collisions don't matter, these are link-local addresses - so only the two addresses of a single tunnel must not collide.

Making this symmetrical by deriving the server address using the same algorithm (which I'd really like) requires changing the server side anyways.

Reply:
I think, collisions actually do matter.

IIRC, the vxlan code in linux can not handle link-local addresses, which do not have "a kernel wide" route. At least in *wireguard-vxlan-glue* we add such a route ([see here](https://github.com/freifunkh/wireguard-vxlan-glue/blob/master/netlink.py#L165-L171)). Also the vxlan.sh proto establishes a kernel wide route for `fe80::1` with the configuration from this PR:

```
root@crashtester:~# ip -6 route show fe80::1
fe80::1 dev vpn proto static metric 1024 pref medium
```

As we need "kernel wide routes" for the link-local addresses, the link-local addresses must be unique.

Reply:
It's also the IP which is included in the `allowed-ips` of wireguard so it _must_ not collide otherwise wireguard doesn't work.

Reply:
Ah, it is very unfortunate that WG requires you to care about the addresses in the tunnel, and doesn't just create an interface per peer.

The VXLAN issue sounds weird, I'll look into that. Such a route should usually not make a difference.

Reply:
It's actually the neat thing that we don't require a gazillion interfaces but just one for all peers :). Which makes handling the server side way easier. Because you can pre-create all interfaces via systemd, ifupdown or whatever and just add peers via netlink. So no dynamic interface creation thus no unmanaged devices.

Reply:
Yeah, I can see benefits for both configurations - OpenVPN does the same, requiring to configure addresses to be routed through the tunnel on the OpenVPN side in TUN mode.

I usually prefer to have this information in the Linux routing table, requiring a separate interface for each L3 tunnel (and if I get around to adding WG support to fastd, that is likely how I'm going to implement it, assigning a separate WG instance and UDP port to each tunnel).

That being said, I'm okay with the current implementation for wgpeerselector in Gluon, but the md5sum still looks weird...

Reply:
I would like to keep the md5sum avoid ugly migrations of existing devices. Is this acceptable?

Reply:
Yes, let's keep it - we have other places in Gluon where we use md5sum to derive MAC addresses.